react-exercises/use-reducer/src/hooks/useReducer.js

```diff
@@ -0,0 +1,34 @@
+import { useState } from 'react';
+
+const dispatcher = {
+    reducer: null,
+    setState: null,
+    state: null,
+    dispatch: (...args) => {
+        const updatedState = dispatcher.reducer(dispatcher.state, ...args);
+
+        dispatcher.setState(updatedState);
+    }, 
+    useReducer(...args) {
+        const [reducer, initialValue, init] = args;
+
+        if(typeof reducer !== 'function'){
+            throw new TypeError('reducer must be function.');
+        }
+        
+        
+        const [state, setState] = useState(() => {
+            return typeof init === 'function' ? init(initialValue) : initialValue;
+        });
+        
+        dispatcher.reducer = reducer;
+        dispatcher.state = state;
+        dispatcher.setState = setState;
+
+        return [state, dispatcher.dispatch]
+    }
+}
+
+export default function useReducer(reducer, initialValue, init) {
+    return dispatcher.useReducer(reducer, initialValue, init);
```

Comment:
Any reason why you put implementation inside an object called dispatcher?

Reply:
React documentation mention following note:
React guarantees that dispatch function identity is stable and won’t change on re-renders.
So, on every rerender the reference of dispatch should not change.
So that, I wrap implmentation inside an object.

Reply:
```js
    const [dispatcher] = useState(() => {
        return getDispatcher();
    });
```

In this, the dispathcer is created again and again.

> React guarantees that dispatch function identity is stable and won’t change on re-renders.
Have you checked this in your implementation? How did you do that? Write a test to assert this, that'll be better.

Reply:
@ArfatSalman , You are right, it gave me different values on every render.
Also, Why react useReducer gave me different reference of 'dispatch' function on every render?
They mention in doc that dispatch function identity is stable and won’t change on re-renders.